### PR TITLE
[Routing] Replace previously added preg_match with preg_replace_callback

### PR DESCRIPTION
**Pull request by arjenm**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | refactor |
| New feature? | performance fix, that also deprecates little used and dangerous parameter |
| BC breaks? | yes (deprecates parameter that should be very little used) |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | as discussed in #18257 and #18379 |
| License | MIT |
| Doc PR | See discussions of as discussed in #18257 and #18379 |

This change was originally provided via #18379, but it was considered a BC-breakage due to the deprecation of the decodedChars-parameter on the UrlGenerator-class.

Please note that while that parameter could be seen as an extension point, it was also prone to errors in routing: by altering that array the UrlGenerator's output was not guaranteed to be RFC-safe or to match the UrlMatcher's expectations. I.e. one could set %2F to be converted to \ rather than /, completely breaking url-matching.

This PR is to remove the array, it was initially intended for its performance gain. But it also yields safer url generation (see before) and has more logical code. The old code percent-encoded (almost) all chars, but reverted some to their original character.

By replacing it with a regexp, we can inverse the behaviour of the 'rawurlencode, than decode "wanted chars" back via strtr' into 'only rawurlencode the unwanted chars'. This gives a small (in absolute terms) performance gain, since strtr is fairly slow. Due to its high relative gain it starts to adds up when you generate many urls; in the tests quoted in https://github.com/symfony/symfony/pull/18257#issuecomment-203878928 it'd gain about 0,5ms for every 100 url's.
While that is a minor absolute number, for pages (like ours) that generally complete within 50ms, that's still a 1% gain by simply having faster url generation.

See this test script as well: http://achelois.tweakers.net/~acm/urlrawencode_or_preg_replace.php
The not-empty test of the BC-addition will probably counter part of that gain, though.

### CODE REVIEW

**Comment by fabpot** [comment]
ping @Tobion

**Comment by Tobion** [comment]
http://achelois.tweakers.net/~acm/urlrawencode_or_preg_replace.php is not reachable.

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php [hunk 1]
```diff
@@ -28,6 +28,30 @@
 class UrlGenerator implements UrlGeneratorInterface, ConfigurableRequirementsInterface
 {
     /**
+     * This regexp matches all characters that should be percent encoded in paths for url's generated by this class.
```
Comment (Tobion) [inline src/Symfony/Component/Routing/Generator/UrlGenerator.php]: URLs

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php [hunk 2]
```diff
@@ -28,6 +28,30 @@
...
+     * - The slash '/' can be used to designate a hierarchical structure and we allow using it with this meaning,
+     *   some webservers don't allow the slash in encoded form in the path for security reasons anyway
+     *   see http://stackoverflow.com/questions/4069002/http-400-if-2f-part-of-get-url-in-jboss
+     * - The '@' and ':' are general delimiters in the URI specification but have only special meaning in the authority component,
+     *   so they can safely be used in the path in unencoded form.
+     * - The ';', ',', '=', '+', '!', '*', '|' are only sub-delimiters that have no predefined meaning and can therefore be used literally
+     *   so URI producing applications can use these chars to delimit subcomponents in a path segment without being encoded for better readability.
+     *
+     * The regexp is an inverse of the characters above. It allows to only call rawurlencode() for specific charaters,
+     * thus optimizing scenario's where there are no or very few such characters in urls.
```
Comment (Tobion) [inline src/Symfony/Component/Routing/Generator/UrlGenerator.php]: scenarios

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php [hunk 3]
```diff
@@ -48,37 +72,11 @@ class UrlGenerator implements UrlGeneratorInterface, ConfigurableRequirementsInt
...
-        '%3D' => '=',
-        '%2B' => '+',
-        '%21' => '!',
-        '%2A' => '*',
-        '%7C' => '|',
-    );
-
-    /**
-     * @var string This regexp matches all characters that are not or should not be encoded by rawurlencode (see list in array above).
+     * @deprecated This array should not be used anymore. Changing it can result in url's that are incompatible with Symfony's url-matching, browsers and/or server software.
```
Comment (Tobion) [inline src/Symfony/Component/Routing/Generator/UrlGenerator.php]: should just say `@deprecated since version 3.1, to be removed in 4.0.`

**Comment by Tobion** [comment]
> it was also prone to errors in routing: by altering that array the UrlGenerator's output was not guaranteed to be RFC-safe

This argument is irrelevant as every extension point can be misused.

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php [hunk 4]
```diff
@@ -187,9 +185,17 @@ protected function doGenerate($variables, $defaults, $requirements, $tokens, $pa
 
         if ('' === $url) {
             $url = '/';
-        } elseif (preg_match($this->urlEncodingSkipRegexp, $url)) {
-            // the context base URL is already encoded (see Symfony\Component\HttpFoundation\Request)
+        } elseif (!empty($this->decodedChars)) {
+            // Support old style of character replacement, for when someone changed the decodedChars-array,
```
Comment (Tobion) [inline src/Symfony/Component/Routing/Generator/UrlGenerator.php]: please remove this comment as it's obvious given the deprecation

File: src/Symfony/Component/Routing/Generator/UrlGenerator.php [hunk 5]
```diff
@@ -28,6 +28,30 @@
...
+     *   so they can safely be used in the path in unencoded form.
+     * - The ';', ',', '=', '+', '!', '*', '|' are only sub-delimiters that have no predefined meaning and can therefore be used literally
+     *   so URI producing applications can use these chars to delimit subcomponents in a path segment without being encoded for better readability.
+     *
+     * The regexp is an inverse of the characters above. It allows to only call rawurlencode() for specific charaters,
+     * thus optimizing scenario's where there are no or very few such characters in urls.
+     *
+     * @internal
+     */
+    const PATH_UNSAFE_CHARACTER_REGEXP = '#[^-.~a-zA-Z0-9_/@:;,=+!*|]+#';
```
Comment (Tobion) [inline src/Symfony/Component/Routing/Generator/UrlGenerator.php]: `'#[^-.~a-zA-Z0-9_/@:;,=+!*|]++#'<´ to prevent backtracking

**Comment by arjenm** [comment]
I fixed that server, so achelois.tweakers.net should be available again.

**Comment by arjenm** [comment]
@Tobion I think I got all your comments

**Comment by fabpot** [comment]
Reverted #18257 instead as @Tobion was against the change in the first place and this PR introduces yet some other big changes that we cannot make in 2.7 where the original PR was merged. Sorry for the back and forth on this topic.

**Comment by Tobion** [comment]
Btw, I also ran your benchmarks against PHP 7 and it showed a performance decrease. So IMO there is no consistently better way than the current approach.

**Pull request closed without merging**